clients/include/device_batch_vector.hpp

```diff
@@ -0,0 +1,236 @@
+
+    //!
+    //! @brief Constructor.
+    //! @param batch_count The number of vectors.
+    //! @param size_vector The size of each vectors.
+    //!
+    explicit device_batch_vector(rocblas_int batch_count, size_t size_vector)
+        : m_batch_count(batch_count)
+        , m_size_vector(size_vector)
+        , d_vector<T, PAD, U>(size_vector)
+    {
+        if(false == this->try_initialize_memory())
+        {
+            this->free_memory();
+        }
+    };
+
+    //!
+    //! @brief Destructor.
+    //!
+    inline ~device_batch_vector() noexcept
+    {
+        this->free_memory();
+    }
+
+    //!
+    //! @brief Returns the size of the vectors.
+    //!
+    inline size_t size() const
+    {
+        return this->m_size_vector;
+    };
+
+    //!
+    //! @brief Returns the value of batch_count.
+    //!
+    inline rocblas_int batch_count() const noexcept
+    {
+        return this->m_batch_count;
+    };
+
+    //!
+    //! @brief Access to device data.
+    //! @return Pointer to the device data.
+    //!
+    inline T** ptr_on_device() noexcept
+    {
+        return this->m_device_data;
+    }
+
+    //!
+    //! @brief Const access to device data.
+    //! @return Const pointer to the device data.
+    //!
+    inline const T* const* ptr_on_device() const noexcept
+    {
+        return this->m_device_data;
+    }
+
+    //!
+    //! @brief Random access.
+    //! @param batch_index The batch index.
+    //! @return Pointer to the array on device.
+    //!
+    inline T* operator[](rocblas_int batch_index) noexcept
+    {
+        assert(0 <= batch_index && this->m_batch_count > batch_index);
+
+        return this->m_data[batch_index];
+    }
+
+    //!
+    //! @brief Constant random access.
+    //! @param batch_index The batch index.
+    //! @return Constant pointer to the array on device.
+    //!
+    inline const T* operator[](rocblas_int batch_index) const noexcept
+    {
+        assert(0 <= batch_index && this->m_batch_count > batch_index);
+
+        return this->m_data[batch_index];
+    }
+
+    //!
+    //! @brief Const cast of the data on host.
+    //!
+    inline operator const T* const*() const noexcept
+    {
+        return this->m_data;
+    };
+
+    //!
+    //! @brief Cast of the data on host.
+    //!
+    inline operator T**() noexcept
+    {
+        return this->m_data;
+    }
+
+    //!
+    //! @brief Tell whether ressources allocation failed.
+    //!
+    inline explicit operator bool() const noexcept
+    {
+        return this->m_data != nullptr;
+    }
+
+    //!
+    //! @brief Copy from a host batched vector.
+    //! @param that The host_batch_vector to copy.
+    //!
+    inline hipError_t transfer_from(const host_batch_vector<T>& that) noexcept
+    {
+        //
+        // Check sizes.
+        //
+        if((this->batch_count() != that.batch_count()) || (this->size() != that.size()))
+        {
+            return hipErrorInvalidContext;
```

Comment:
Is this consistent with hip interpretation?  Maybe hipErrorUnknown ?   Or hipErrorInvalidValue which is what  would happen if one array is extended with null_ptr to match.

Reply:
I found this one meaningful, but I'm open to any other.

Reply:
We should define our own `rocblas_status_*` error codes (or use ones which exist), instead of returning `hipError_t`. There are macros for converting from `hipError_t` to `rocblas_status` if we have to detect errors in HIP calls.

Reply:
Currently, CHECK_HIP_ERROR is widely used in testings, especially from HIP calls (transfer).  So I am trying to keep tight with the existing 'philosophy'. But I'm willing to change the error type.

Reply:
`CHECK_HIP_ERROR()` checks the HIP return type passed as its argument, and if it is not `hipSuccess`, prints an error message, and calls `exit(1)`. I do not see how this is similar to this context, where a different return value is returned if there is a HIP error.

HIP is separate from rocBLAS, and there is no need to compare them, or to try to make them behave similarly.

All rocBLAS APIs should return rocBLAS return values, and by extension, rocBLAS tests should return rocBLAS status codes instead of HIP status codes. If there is no return value, and an exceptional condition needs to be indicated, then it depends on the context what the correct action should be.

Note that Google Tests do not detect C++ exceptions, and `exit()` and `abort()` should only be used for fatal error conditions, not just failing tests.

If a function you're writing can indicate an error in its return value, then it should do so where possible, even if it means returning nullptr.

Reply:
CHECK_HIP_ERROR is used in the tests to check the validity of hip operations.
It seems natural to return a hip error if a hip operation is used.
I don't see where the problem is. If this is about the specificity of the hipErrorInvalidContext, this is the one I found with the closest meaning.

Is this a rocblas api? No, this is utility classes for testing.
Question: Can we converge to a solution ?

- 1/ return hip error?
- 2/ return rocblas error ? That would mean more work when refactoring of tests (by not using CHECK_HIP_ERROR)